src/client/datascience/kernel-launcher/localKernelSpecFinderBase.ts

```diff
@@ -0,0 +1,196 @@
+        cancelToken?: CancellationToken
+    ): Promise<IJupyterKernelSpec | undefined> {
+        // If we have not already loaded this kernel spec, then load it
+        if (!this.pathToKernelSpec.has(specPath)) {
+            this.pathToKernelSpec.set(specPath, this.loadKernelSpec(specPath, interpreter, cancelToken));
+        }
+
+        // ! as the has and set above verify that we have a return here
+        return this.pathToKernelSpec.get(specPath)!.then((value) => {
+            if (value) {
+                return value;
+            }
+
+            // If we failed to get a kernelspec full path from our cache and loaded list
+            this.pathToKernelSpec.delete(specPath);
+            this.cache = this.cache?.filter((itempath) => itempath.kernelSpecFile !== specPath);
+            return undefined;
+        });
+    }
+
+    /**
+     * Load kernelspec json from disk
+     */
+    private async loadKernelSpec(
+        specPath: string,
+        interpreter?: PythonEnvironment,
+        cancelToken?: CancellationToken
+    ): Promise<IJupyterKernelSpec | undefined> {
+        let kernelJson: ReadWrite<IJupyterKernelSpec>;
+        try {
+            traceInfo(`Loading kernelspec from ${specPath} for ${interpreter?.path}`);
+            kernelJson = JSON.parse(await this.fs.readLocalFile(specPath));
+        } catch {
+            traceError(`Failed to parse kernelspec ${specPath}`);
+            return undefined;
+        }
+        if (cancelToken?.isCancellationRequested) {
+            return undefined;
+        }
+
+        // Special case. If we have an interpreter path this means this spec file came
+        // from an interpreter location (like a conda environment). Modify the name to make sure it fits
+        // the kernel instead
+        kernelJson.name = interpreter ? getInterpreterKernelSpecName(interpreter) : kernelJson.name;
+
+        // Update the display name too if we have an interpreter.
+        kernelJson.display_name =
+            kernelJson.language === PYTHON_LANGUAGE
+                ? interpreter?.displayName || kernelJson.display_name
+                : kernelJson.display_name;
+
+        const kernelSpec: IJupyterKernelSpec = new JupyterKernelSpec(
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            kernelJson as any,
+            specPath,
+            // Interpreter information may be saved in the metadata (if this is a kernel spec created/registered by us).
+            interpreter?.path || kernelJson?.metadata?.interpreter?.path
+        );
+
+        // Some registered kernel specs do not have a name, in this case use the last part of the path
+        kernelSpec.name = kernelJson?.name || path.basename(path.dirname(specPath));
+
+        // Possible user deleted the underlying kernel.
+        const interpreterPath = interpreter?.path || kernelJson?.metadata?.interpreter?.path;
+        if (
+            isKernelRegisteredByUs(kernelSpec) &&
```

Comment:
I think this was from a previous change (a bit hard to tell since there is all the new code), but why would this be gated on isKernelRegisteredByUs? If there is an interpreter path and it doesn't exist do we want the kernel spec?

Reply:
Yes, that's better, fixed.